### PR TITLE
Add Rotation Initialization for Objects

### PR DESCRIPTION
**Pull request by zhx06**
## Summary
Add random yaw initialization for placed objects

## Detailed description
- Add `--random_yaw_init` flag for per-object random z-rotation.
- Add conservative rotated-AABB support so collision checks account for yaw.
- Add related tests.

### CODE REVIEW

File: isaaclab_arena/environments/relation_solver_interface.py [hunk 1]
```diff
@@ -35,6 +36,8 @@ def solve_and_apply_relation_placement(
         resolve_on_reset: Optional override for whether to draw fresh layouts from
             the placement pool on reset. When ``False``, fixed per-environment
             initial poses are applied immediately.
+        random_yaw_init: If True, give each non-anchor object a random fixed yaw about Z
```
Comment (xyao-nv) [inline isaaclab_arena/environments/relation_solver_interface.py]: Docstring is lil jorgan-y, to make it clean by
`If True, randomly rotates non-anchor objects around the vertical (Z) axis at startup to add visual variety to the scene.`

File: isaaclab_arena/utils/pose.py [hunk 2]
```diff
@@ -52,6 +53,19 @@ def multiply(self, other: "Pose") -> "Pose":
         return compose_poses(self, other)
 
 
+def rotate_quat_by_yaw(
```
Comment (xyao-nv) [inline isaaclab_arena/utils/pose.py:61]: Can you add a unit test for this math func?

File: isaaclab_arena/utils/pose.py [hunk 3]
```diff
@@ -52,6 +53,19 @@ def multiply(self, other: "Pose") -> "Pose":
...
+    base_xyzw: tuple[float, float, float, float], yaw_rad: float
+) -> tuple[float, float, float, float]:
+    """Rotate base_xyzw (xyzw) by an extra yaw about Z. Returns base unchanged when yaw is 0."""
+    if yaw_rad == 0.0:
```
Comment (xyao-nv) [inline isaaclab_arena/utils/pose.py:66]: How about wrapping yaw_rad to [-pi, pi] to avoid unnecessary compute and potential errors caused by precision?

File: isaaclab_arena/relations/placement_result.py [hunk 4]
```diff
@@ -28,6 +28,9 @@ class PlacementResult:
     attempts: int
     """Number of attempts made."""
 
+    orientations: dict[ObjectBase, float] = field(default_factory=dict)
+    """Per-object fixed yaw (radians about Z) for this layout. Empty when unrotated."""
```
Comment (xyao-nv) [inline isaaclab_arena/relations/placement_result.py]: It will be more informative if you clarify w.r.t to what frame.
> ``fixed yaw``

File: isaaclab_arena/relations/relation_solver_state.py [hunk 5]
```diff
@@ -93,6 +103,19 @@ def __init__(
...
 
+        # Precompute each object's local bbox on the solver device. Rotate to the conservative
+        # enclosing box (shape (batch_size, 3)) only when some per-env yaw is non-zero;
+        # otherwise keep the base (1, 3) box, which the loss math broadcasts across envs.
+        self._local_bboxes: dict[ObjectBase, AxisAlignedBoundingBox] = {}
+        for obj in objects:
+            bbox = obj.get_bounding_box().to(self._device)
+            if initial_orientations is not None:
+                yaws = [d.get(obj, 0.0) for d in initial_orientations]
+                if any(yaw != 0.0 for yaw in yaws):
```
Comment (xyao-nv) [inline isaaclab_arena/relations/relation_solver_state.py]: again, I strongly encourage wrapping yaw before branching and applying the transformation

File: isaaclab_arena/relations/placement_events.py [hunk 6]
```diff
@@ -50,17 +50,18 @@ def solve_and_place_objects(
...
 
     anchor_objects_set = set(get_anchor_objects(objects))
-    rotations = {obj: get_rotation_xyzw(obj) for obj in objects if obj not in anchor_objects_set}
 
     zero_velocity = torch.zeros(1, 6, device=env.device)
     for local_idx, cur_env in enumerate(env_ids.tolist()):
         env_id_tensor = torch.tensor([cur_env], device=env.device)
-        positions = layouts_per_env[local_idx].positions
+        result = layouts_per_env[local_idx]
+        positions = result.positions
```
Comment (xyao-nv) [inline isaaclab_arena/relations/placement_events.py]: Is it needed? You are only reusing it once next line.
`results.orientations` stays its own tho.

File: isaaclab_arena/relations/object_placer.py [hunk 7]
```diff
@@ -301,9 +329,24 @@ def _sample_axis_position(
...
+        self,
+        obj: ObjectBase,
+        position: tuple[float, float, float],
+        orientations: dict[ObjectBase, float] | None,
+    ) -> AxisAlignedBoundingBox:
+        """World-space bbox for obj at position, applying its sampled yaw (0 if absent)
+        via the same conservative enclosing AABB the solver uses, so validation matches solving."""
+        bbox = obj.get_bounding_box()
+        yaw = (orientations or {}).get(obj, 0.0)
+        if yaw != 0.0:
```
Comment (xyao-nv) [inline isaaclab_arena/relations/object_placer.py]: wrapping

File: isaaclab_arena/utils/pose.py (hunk 3 above)
Comment (xyao-nv) [reply isaaclab_arena/utils/pose.py:66]: I see, in `_generate_initial_orientations()` it outs [-pi, pi].
But for this helper func alone, it could be reused with user defined `yaw_rad`.

File: isaaclab_arena/relations/relation_solver_state.py (hunk 5 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relation_solver_state.py]: Either you can ensure yaw is only provided thru `_generate_initial_orientations()`, so [-pi, pi] is enforced.
Or you can add a 'if not wrapper, wrap it' to be safe.

File: isaaclab_arena/relations/object_placer.py [hunk 8]
```diff
@@ -423,19 +474,24 @@ def _apply_positions(
...
+                rotation_xyzw = rotate_quat_by_yaw(base_xyzw, orientations_per_env[0].get(obj, 0.0))
                 random_marker = self._get_random_around_solution(obj)
                 if random_marker is not None:
                     obj.set_initial_pose(random_marker.to_pose_range_centered_at(pos, rotation_xyzw=rotation_xyzw))
                 else:
                     obj.set_initial_pose(Pose(position_xyz=pos, rotation_xyzw=rotation_xyzw))
             else:
                 poses = [
-                    Pose(position_xyz=positions_per_env[env_idx][obj], rotation_xyzw=rotation_xyzw)
+                    Pose(
```
Comment (xyao-nv) [inline isaaclab_arena/relations/object_placer.py:650]: Shall this func be updated to `_apply_pose()`

File: isaaclab_arena/relations/relation_solver_state.py [hunk 9]
```diff
@@ -93,6 +103,19 @@ def __init__(
             self._opt_idx_tensor = None
             self._optimizable_positions = None
 
+        # Precompute each object's local bbox on the solver device. Rotate to the conservative
```
Comment (xyao-nv) [inline isaaclab_arena/relations/relation_solver_state.py]:
```
# Precompute original bounding boxes on the GPU/device.
# If an object is rotated (yaw != 0) in any environment, calculate a larger,
# unrotated "enclosing" box to fit it, creating a separate box per environment.
# Otherwise, keep a single base box and let the loss function broadcast it to save memory.
```

^ Is my understanding correct?

File: isaaclab_arena/cli/isaaclab_arena_cli.py [hunk 10]
```diff
@@ -75,6 +75,15 @@ def add_isaaclab_arena_cli_args(parser: argparse.ArgumentParser) -> None:
             " layout."
         ),
     )
+    arena_group.add_argument(
+        "--random_yaw_init",
+        action="store_true",
+        default=False,
+        help=(
+            "Initialize non-anchor placeable objects with a random yaw about the up (Z) axis for scene variety. "
```
Comment (xyao-nv) [inline isaaclab_arena/cli/isaaclab_arena_cli.py]:
```
help=(
        "Randomly rotate objects (except anchors) around the Z-axis for scene variety. "
        "Collisions use a larger enclosing box; the solver won't optimize this rotation."
    ),
```

File: isaaclab_arena/environments/relation_solver_interface.py (hunk 1 above)
Comment (zhx06) [reply isaaclab_arena/environments/relation_solver_interface.py]: Done. Docstring updated as suggested.

File: isaaclab_arena/utils/pose.py (hunk 2 above)
Comment (zhx06) [reply isaaclab_arena/utils/pose.py:61]: Unit tests added.

File: isaaclab_arena/utils/pose.py (hunk 3 above)
Comment (zhx06) [reply isaaclab_arena/utils/pose.py:66]: Done. It is wrapped inside the helper.

File: isaaclab_arena/relations/relation_solver_state.py (hunk 5 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver_state.py]: Done. Defensive wrap is applied.

File: isaaclab_arena/relations/object_placer.py (hunk 7 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py]: Done. This has been replaced after the rebase.

File: isaaclab_arena/relations/placement_events.py (hunk 6 above)
Comment (zhx06) [reply isaaclab_arena/relations/placement_events.py]: Done. The redundant `positions` are removed.

File: isaaclab_arena/relations/object_placer.py (hunk 8 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py:650]: Done. It is renamed as `_apply_poses`.

File: isaaclab_arena/relations/relation_solver_state.py (hunk 9 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver_state.py]: Yes, it is correct.

File: isaaclab_arena/cli/isaaclab_arena_cli.py (hunk 10 above)
Comment (zhx06) [reply isaaclab_arena/cli/isaaclab_arena_cli.py]: Done. text has been updated as suggested.

File: isaaclab_arena/relations/placement_result.py (hunk 4 above)
Comment (zhx06) [reply isaaclab_arena/relations/placement_result.py]: Done. It is clarified that it's in the world frame.

File: isaaclab_arena/cli/isaaclab_arena_cli.py [hunk 11]
```diff
@@ -75,6 +75,15 @@ def add_isaaclab_arena_cli_args(parser: argparse.ArgumentParser) -> None:
             " layout."
         ),
     )
+    arena_group.add_argument(
+        "--random_yaw_init",
+        action="store_true",
+        default=False,
+        help=(
+            "Randomly rotate objects (except anchors) around the Z-axis for scene variety. "
+            "Collisions use a larger enclosing box; the solver won't optimize this rotation."
```
Comment (alexmillane) [inline isaaclab_arena/cli/isaaclab_arena_cli.py]: Should we add that this parameter only takes effect when using the object placement? I.e. if users are manually placing objects they won't be rotated.

File: isaaclab_arena/environments/relation_solver_interface.py [hunk 12]
```diff
@@ -126,7 +130,7 @@ def _apply_dynamic_spawn_pose(
         object_cfg = getattr(obj, "object_cfg", None)
         assert object_cfg is not None, f"Object '{obj.name}' must have object_cfg initialized before placement."
         object_cfg.init_state.pos = pos
-        object_cfg.init_state.rot = get_rotation_xyzw(obj)
+        object_cfg.init_state.rot = rotate_quat_by_yaw(get_rotation_xyzw(obj), layout.orientations.get(obj, 0.0))
```
Comment (alexmillane) [inline isaaclab_arena/environments/relation_solver_interface.py]: Split over multiple lines

```
pos = layout.positions.get(obj)
if pos is None:
    continue
yaw = layout.orientations.get(obj, 0.0)
rot = rotate_quat_by_yaw(get_rotation_xyzw(obj), yaw)
object_cfg = getattr(obj, "object_cfg", None)
assert object_cfg is not None, f"Object '{obj.name}' must have object_cfg initialized before placement."
object_cfg.init_state.pos = pos
object_cfg.init_state.rot = rot
...
```

File: isaaclab_arena/environments/relation_solver_interface.py [hunk 13]
```diff
@@ -149,14 +153,15 @@ def _apply_static_initial_poses(
...
-        rotation_xyzw = get_rotation_xyzw(obj)
+        base_rotation_xyzw = get_rotation_xyzw(obj)
         poses = []
         missing_envs: list[int] = []
         for env_idx in range(num_envs):
             pos = layouts[env_idx].positions.get(obj)
             if pos is None:
                 missing_envs.append(env_idx)
             else:
+                rotation_xyzw = rotate_quat_by_yaw(base_rotation_xyzw, layouts[env_idx].orientations.get(obj, 0.0))
```
Comment (alexmillane) [inline isaaclab_arena/environments/relation_solver_interface.py]: See comment above. Split rotation over multiple lines.

File: isaaclab_arena/relations/object_placer.py [hunk 14]
```diff
@@ -313,6 +328,87 @@ def _get_world_bbox_for_init(
...
+        row_idx: int,
+    ) -> dict[ObjectBase, AxisAlignedBoundingBox]:
+        """Slice one candidate's bboxes (each (1, 3)) out of the stacked (num_candidates, 3) boxes."""
+        return {
+            obj: AxisAlignedBoundingBox(
+                min_point=bbox.min_point[row_idx : row_idx + 1],
+                max_point=bbox.max_point[row_idx : row_idx + 1],
+            )
+            for obj, bbox in bboxes.items()
+        }
```
Comment (alexmillane) [inline isaaclab_arena/relations/object_placer.py]: Suggestion to move to this slicing into `AxisAlignedBoundingBox`. One idea would be to have this as an accessor function `AxisAlignedBoundingBox[idx]`.

The method could then be:

```
def _get_bounding_boxes_for_candidate_index(bounding_boxes: AxisAlignedBoundingBox,idx: int):
    return {obj: bbox[idx] for bbox in bboxes.items()}
```

File: isaaclab_arena/relations/object_placer.py [hunk 15]
```diff
@@ -313,6 +328,87 @@ def _get_world_bbox_for_init(
...
+        Empty dict (no RNG consumed) when random_yaw_init is off; anchors are never rotated.
+        """
+        if not self.params.random_yaw_init:
+            return {}
+        orientations: dict[ObjectBase, float] = {}
+        for obj in objects:
+            if obj in anchor_objects:
+                continue
+            u = torch.rand(1, generator=generator).item()
+            orientations[obj] = (2.0 * u - 1.0) * math.pi  # uniform in [-pi, pi)
```
Comment (alexmillane) [inline isaaclab_arena/relations/object_placer.py]: Suggestion to move to `isaaclab_arena/utils/rand.py` as `get_random_rotation`

File: isaaclab_arena/relations/object_placer.py [hunk 16]
```diff
@@ -196,12 +202,19 @@ def _place_ranked(
...
+        # Bake each candidate's yaw into a conservative enclosing bbox; no-op when yaw is disabled.
+        # The solver and validation then treat the rotated object as an axis-aligned box.
+        candidate_bboxes = self._rotate_candidate_bboxes(objects, candidate_bboxes, orientations_per_candidate)
 
         all_positions = self._solver.solve(objects, initial_positions, env_bboxes=candidate_bboxes)
         assert self._solver.last_loss_per_env is not None
         all_losses: list[float] = self._solver.last_loss_per_env.cpu().tolist()
         all_validations = [
-            self._validate_placement(positions, per_env_bboxes[candidate_idx // candidates_per_env])
+            self._validate_placement(positions, self._bbox_row(candidate_bboxes, candidate_idx))
```
Comment (alexmillane) [inline isaaclab_arena/relations/object_placer.py]: What's meant by `_bbox_row`....

_Edit: I checked below. Please see recommendations there._

File: isaaclab_arena/relations/object_placer.py [hunk 17]
```diff
@@ -313,6 +328,87 @@ def _get_world_bbox_for_init(
...
+                    for c in range(num_candidates)
+                ]
+                if any(yaw != 0.0 for yaw in yaws):
+                    yaw_tensor = torch.tensor(yaws, dtype=torch.float32, device=bbox.min_point.device)
+                    bbox = bbox.rotated_around_z(yaw_tensor)
+            rotated[obj] = bbox
+        return rotated
+
+    @staticmethod
+    def _marker_z_yaw(obj: ObjectBase) -> float:
```
Comment (alexmillane) [inline isaaclab_arena/relations/object_placer.py]: `_get_yaw_from_rotate_around_solution()`

File: isaaclab_arena/relations/object_placer.py [hunk 18]
```diff
@@ -313,6 +328,87 @@ def _get_world_bbox_for_init(
...
+        for obj in objects:
+            bbox = candidate_bboxes[obj]
+            # Only objects that receive a sampled yaw are rotated; anchors never appear here.
+            if any(obj in orientations for orientations in orientations_per_candidate):
+                # Enclose marker_rotation + sampled yaw (the applied pose); both are pure-Z.
+                marker_yaw = ObjectPlacer._marker_z_yaw(obj)
+                yaws = [
+                    wrap_angle_to_pi(orientations_per_candidate[c].get(obj, 0.0) + marker_yaw)
+                    for c in range(num_candidates)
+                ]
```
Comment (alexmillane) [inline isaaclab_arena/relations/object_placer.py:376]: What's the intended behaviour here? Is my interprettation correct that we're adding a random rotation to the yaw specified by the marker? What's the effect of that?

**Review by alexmillane — APPROVED** [review]
Looks good.

Made some small nits across the code. Feel free to merge once addressed.

File: isaaclab_arena/cli/isaaclab_arena_cli.py (hunk 11 above)
Comment (zhx06) [reply isaaclab_arena/cli/isaaclab_arena_cli.py]: Addressed. Comments have been added to the text in help.

File: isaaclab_arena/environments/relation_solver_interface.py (hunk 12 above)
Comment (zhx06) [reply isaaclab_arena/environments/relation_solver_interface.py]: Done. The code now follows the suggested pattern.

File: isaaclab_arena/environments/relation_solver_interface.py (hunk 13 above)
Comment (zhx06) [reply isaaclab_arena/environments/relation_solver_interface.py]: Done. Same as above.

File: isaaclab_arena/relations/object_placer.py (hunk 14 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py]: Done. `AxisAlignedBoundingBox.__getitem__(idx)` added

File: isaaclab_arena/relations/object_placer.py (hunk 17 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py]: Done. Function renamed.

File: isaaclab_arena/relations/object_placer.py (hunk 16 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py]: Done. `_bbox_row` is now `_get_bounding_boxes_for_candidate_index`

File: isaaclab_arena/relations/object_placer.py (hunk 15 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py]: Done. The function is now in `isaaclab_arena/utils/random.py `.

File: isaaclab_arena/relations/object_placer.py (hunk 18 above)
Comment (zhx06) [reply isaaclab_arena/relations/object_placer.py:376]: Yes, your interpretation is correct. The object's final yaw is `marker_yaw` + `random_yaw`, which means we treat the marker as a starting offset.
